diff -ur termite.orig/termite.cc termite/termite.cc
--- termite.orig/termite.cc	2020-04-24 17:57:12.341506512 +0900
+++ termite/termite.cc	2020-04-24 17:58:29.768518090 +0900
@@ -779,23 +779,13 @@
 static void increase_font_scale(VteTerminal *vte) {
     gdouble scale = vte_terminal_get_font_scale(vte);
 
-    for (auto it = zoom_factors.begin(); it != zoom_factors.end(); ++it) {
-        if ((*it - scale) > 1e-6) {
-            vte_terminal_set_font_scale(vte, *it);
-            return;
-        }
-    }
+    vte_terminal_set_font_scale(vte, scale * 1.1);
 }
 
 static void decrease_font_scale(VteTerminal *vte) {
     gdouble scale = vte_terminal_get_font_scale(vte);
 
-    for (auto it = zoom_factors.rbegin(); it != zoom_factors.rend(); ++it) {
-        if ((scale - *it) > 1e-6) {
-            vte_terminal_set_font_scale(vte, *it);
-            return;
-        }
-    }
+    vte_terminal_set_font_scale(vte, scale / 1.1);
 }
 
 gboolean window_state_cb(GtkWindow *, GdkEventWindowState *event, keybind_info *info) {
