diff -aur rsnapshot-1.4.1.pristine/rsnapshot-program.pl rsnapshot-1.4.1.new/rsnapshot-program.pl
--- rsnapshot-1.4.1.pristine/rsnapshot-program.pl	2015-11-08 17:41:24.998161558 +0900
+++ rsnapshot-1.4.1.new/rsnapshot-program.pl	2015-11-08 17:58:19.544076770 +0900
@@ -155,6 +155,7 @@
 my $default_rsync_long_args  = '--delete --numeric-ids --relative --delete-excluded';
 my $default_ssh_args         = undef;
 my $default_du_args          = '-csh';
+my $default_mount_args       = '';
 
 # set default for use_lazy_deletes
 my $use_lazy_deletes = 0;    # do not delete the oldest archive until after backup
@@ -1380,6 +1381,13 @@
 			next;
 		}
 
+		# MOUNT ARGS
+		if ($var eq 'mount_args') {
+			$config_vars{'mount_args'} = $value;
+			$line_syntax_ok = 1;
+			next;
+		}
+
 		# SSH ARGS
 		if ($var eq 'ssh_args') {
 			if (!defined($default_ssh_args) && defined($config_vars{'ssh_args'})) {
@@ -1831,6 +1839,12 @@
 
 			# pass unchecked
 
+			# mount_args
+		}
+		elsif ($name eq 'mount_args') {
+
+			# pass unchecked
+
 			# ssh_args
 		}
 		elsif ($name eq 'ssh_args') {
@@ -3572,6 +3586,7 @@
 	my $ssh_args         = $default_ssh_args;
 	my $rsync_short_args = $default_rsync_short_args;
 	my $rsync_long_args  = $default_rsync_long_args;
+	my $mount_args       = $default_mount_args;
 
 	# other misc variables
 	my @cmd_stack = undef;
@@ -3651,6 +3666,9 @@
 	if (defined($config_vars{'rsync_long_args'})) {
 		$rsync_long_args = $config_vars{'rsync_long_args'};
 	}
+	if (defined($config_vars{'mount_args'})) {
+		$mount_args = $config_vars{'mount_args'};
+	}
 	if (defined($config_vars{'ssh_args'})) {
 		$ssh_args = $config_vars{'ssh_args'};
 	}
@@ -3662,6 +3680,7 @@
 	# quoting - ideally we'd use Text::Balanced or similar, but that's
 	# only relatively recently gone into core
 	my @rsync_long_args_stack = split_long_args_with_quotes('rsync_long_args', $rsync_long_args);
+	my @mount_args_stack = split_long_args_with_quotes('mount_args', $mount_args);
 
 	# create $interval.0/$$bp_ref{'dest'} or .sync/$$bp_ref{'dest'} directory if it doesn't exist
 	# (this may create the .sync dir, which is why we had to check for it above)
@@ -3701,6 +3720,21 @@
 		);
 	}
 
+	# MOUNT ARGS
+	if (defined($$bp_ref{'opts'}) && defined($$bp_ref{'opts'}->{'mount_args'})) {
+		@mount_args_stack = split_long_args_with_quotes('mount_args (for a backup point)',
+			$$bp_ref{'opts'}->{'mount_args'});
+	}
+	if (defined($$bp_ref{'opts'}) && defined($$bp_ref{'opts'}->{'extra_mount_args'})) {
+		push(
+			@mount_args_stack,
+			split_long_args_with_quotes(
+				'extra_mount_args (for a backup point)',
+				$$bp_ref{'opts'}->{'extra_mount_args'}
+			)
+		);
+	}
+
 	# SSH ARGS
 	if (defined($$bp_ref{'opts'}) && defined($$bp_ref{'opts'}->{'ssh_args'})) {
 		$ssh_args = $$bp_ref{'opts'}->{'ssh_args'};
@@ -3778,7 +3812,7 @@
 
 		linux_lvm_snapshot_create(linux_lvm_parseurl($lvm_src));
 		$traps{"linux_lvm_snapshot"} = $lvm_src;
-		linux_lvm_mount(linux_lvm_parseurl($lvm_src));
+		linux_lvm_mount(linux_lvm_parseurl($lvm_src), \@mount_args_stack);
 		$traps{"linux_lvm_mountpoint"} = 1;
 
 		# rewrite src to point to mount path
@@ -4069,7 +4103,7 @@
 
 	my $result = undef;
 
-	my ($linux_lvmvgname, $linux_lvmvolname, $linux_lvmpath) = @_;
+	my ($linux_lvmvgname, $linux_lvmvolname, $linux_lvmpath, $mount_args_stack) = @_;
 	unless (defined($linux_lvmvgname) and defined($linux_lvmvolname) and defined($linux_lvmpath)) {
 		bail("linux_lvm_mount needs 3 parameters!");
 	}
@@ -4078,6 +4112,14 @@
 	my @cmd_stack = ();
 	push(@cmd_stack, split(' ', $config_vars{'linux_lvm_cmd_mount'}));
 
+        if (@$mount_args_stack && (scalar(@$mount_args_stack) > 0)) {
+            foreach my $tmp_arg (@$mount_args_stack) {
+                if (defined($tmp_arg) && ($tmp_arg ne '')) {
+                    push(@cmd_stack, $tmp_arg);
+                }
+            }
+        }
+
 	push(
 		@cmd_stack,
 		join('/',
