--- src/rxvtfont.C.orig	2014-12-13 10:00:23.000000000 +0900
+++ src/rxvtfont.C	2015-07-01 18:58:20.000000000 +0900
@@ -1695,11 +1695,11 @@
     {
       rxvt_font *f = fonts[i];
 
+      if (FROM_UNICODE (f->cs, unicode) == NOCHAR)
+        goto next_font;
+
       if (!f->loaded)
         {
-          if (FROM_UNICODE (f->cs, unicode) == NOCHAR)
-            goto next_font;
-
           if (!realize_font (i))
             goto next_font;
 
