--- fcitx5/src/lib/fcitx/instance.cpp.orig	2025-10-26 15:36:21.596859757 +0900
+++ fcitx5/src/lib/fcitx/instance.cpp	2025-10-26 15:40:46.719904228 +0900
@@ -456,6 +456,9 @@
     if (!q->canTrigger()) {
         return false;
     }
+    if (!ic->inputPanel().empty()) {
+        return false;
+    }
     auto *inputState = ic->propertyFor(&inputStateFactory_);
     return inputState->isActive();
 }
@@ -854,6 +857,10 @@
                         if (origKey.hasModifier()) {
                             inputState->totallyReleased_ = false;
                         }
+                        if ((keyEvent.origKey().states().test(KeyState::NoState) && keyEvent.origKey().sym() == FcitxKey_Escape) ||
+                            (keyEvent.origKey().states().test(KeyState::Ctrl) && keyEvent.origKey().sym() == FcitxKey_bracketleft)) {
+                            break;
+                        }
                         keyEvent.filterAndAccept();
                         return;
                     }
@@ -2264,6 +2271,9 @@
     if (!canTrigger()) {
         return false;
     }
+    if (!ic->inputPanel().empty()) {
+        return false;
+    }
     if (!inputState->isActive()) {
         return true;
     }
