--- fcitx5-5.0.17/src/lib/fcitx/instance.cpp.orig	2022-06-02 03:50:52.000000000 +0900
+++ fcitx5-5.0.17/src/lib/fcitx/instance.cpp	2022-06-04 14:24:25.743909895 +0900
@@ -403,6 +403,9 @@
     if (!q->canTrigger()) {
         return false;
     }
+    if (!ic->inputPanel().empty()) {
+        return false;
+    }
     auto *inputState = ic->propertyFor(&inputStateFactory_);
     return inputState->isActive();
 }
@@ -784,6 +787,12 @@
                         if (origKey.hasModifier()) {
                             inputState->totallyReleased_ = false;
                         }
+
+                        if ((keyEvent.origKey().states().test(KeyState::NoState) && keyEvent.origKey().sym() == FcitxKey_Escape) ||
+                            (keyEvent.origKey().states().test(KeyState::Ctrl) && keyEvent.origKey().sym() == FcitxKey_bracketleft)) {
+                            break;
+                        }
+
                         return keyEvent.filterAndAccept();
                     }
                     idx++;
@@ -1995,6 +2004,9 @@
     if (!canTrigger()) {
         return false;
     }
+    if (!ic->inputPanel().empty()) {
+        return false;
+    }
     if (!inputState->isActive()) {
         return true;
     }
