diff --git a/gtk2/immodule/caret-state-indicator.c b/gtk2/immodule/caret-state-indicator.c
index dfd4537..1d4bd9f 100644
--- a/gtk2/immodule/caret-state-indicator.c
+++ b/gtk2/immodule/caret-state-indicator.c
@@ -136,7 +136,9 @@ caret_state_indicator_new(void)
   gtk_window_set_default_size(GTK_WINDOW(window),
 			      DEFAULT_WINDOW_WIDTH,
 			      DEFAULT_WINDOW_HEIGHT);
+#if !GTK_CHECK_VERSION(3, 0, 0)
   gtk_widget_set_app_paintable(window, TRUE);
+#endif
 
 #if GTK_CHECK_VERSION(2, 90, 0)
   g_signal_connect(window, "draw",
