--- server-client.c.orig	2015-09-27 18:42:08.000000000 +0900
+++ server-client.c	2015-10-24 13:48:25.795601674 +0900
@@ -1235,10 +1235,14 @@
 		return;
 	}
 	if (c->flags & CLIENT_UTF8)
-		c->tty.flags |= TTY_UTF8;
+		c->tty.flags |= TTY_UTF8 | TTY_UTF8_ACS;
 	if (c->flags & CLIENT_256COLOURS)
 		c->tty.term_flags |= TERM_256COLOURS;
 
+	char *env = getenv("TMUX_NO_UTF8_ACS");
+	if (env && strcmp("0", env) != 0)
+		c->tty.flags &= ~TTY_UTF8_ACS;
+
 	tty_resize(&c->tty);
 
 	if (!(c->flags & CLIENT_CONTROL))
--- tmux.h.orig	2017-05-31 18:31:09.560332826 +0900
+++ tmux.h	2017-05-31 18:31:13.870207322 +0900
@@ -1075,6 +1075,7 @@
 #define TTY_OPENED 0x20
 #define TTY_FOCUS 0x40
 #define TTY_BLOCK 0x80
+#define TTY_UTF8_ACS 0x8000
 	int		 flags;
 
 	struct tty_term	*term;
--- tty-acs.c.orig	2014-11-21 05:44:57.000000000 +0900
+++ tty-acs.c	2015-10-22 21:23:38.934933019 +0900
@@ -81,7 +81,7 @@
 	struct tty_acs_entry *entry;
 
 	/* If not a UTF-8 terminal, use the ACS set. */
-	if (tty != NULL && !(tty->flags & TTY_UTF8)) {
+	if (tty != NULL && !(tty->flags & TTY_UTF8_ACS)) {
 		if (tty->term->acs[ch][0] == '\0')
 			return (NULL);
 		return (&tty->term->acs[ch][0]);
--- tty.c.orig	2017-04-25 12:20:21.461598065 +0900
+++ tty.c	2017-04-25 12:22:02.828602056 +0900
@@ -72,7 +72,7 @@
 		    u_int);
 
 #define tty_use_acs(tty) \
-	(tty_term_has((tty)->term, TTYC_ACSC) && !((tty)->flags & TTY_UTF8))
+	(tty_term_has((tty)->term, TTYC_ACSC) && !((tty)->flags & TTY_UTF8_ACS))
 #define tty_use_margin(tty) \
 	((tty)->term_type == TTY_VT420)
 
