--- server-client.c.orig	2015-09-27 18:42:08.000000000 +0900
+++ server-client.c	2015-10-24 13:48:25.795601674 +0900
@@ -1235,10 +1235,14 @@
 		return;
 	}
 	if (c->flags & CLIENT_UTF8)
-		c->tty.flags |= TTY_UTF8;
+		c->tty.flags |= TTY_UTF8 | TTY_UTF8_ACS;
 	if (c->flags & CLIENT_256COLOURS)
 		c->tty.term_flags |= TERM_256COLOURS;
 
+	char *env = getenv("TMUX_NO_UTF8_ACS");
+	if (env && strcmp("0", env) != 0)
+		c->tty.flags &= ~TTY_UTF8_ACS;
+
 	tty_resize(&c->tty);
 
 	if (!(c->flags & CLIENT_CONTROL))
--- tmux.h.orig	2015-09-27 18:51:49.000000000 +0900
+++ tmux.h	2015-10-24 13:26:39.745796432 +0900
@@ -1118,6 +1118,7 @@
 #define TTY_STARTED 0x10
 #define TTY_OPENED 0x20
 #define TTY_FOCUS 0x40
+#define TTY_UTF8_ACS 0x8000
 	int		 flags;
 
 	int		 term_flags;
--- tty-acs.c.orig	2014-11-21 05:44:57.000000000 +0900
+++ tty-acs.c	2015-10-22 21:23:38.934933019 +0900
@@ -81,7 +81,7 @@
 	struct tty_acs_entry *entry;
 
 	/* If not a UTF-8 terminal, use the ACS set. */
-	if (tty != NULL && !(tty->flags & TTY_UTF8)) {
+	if (tty != NULL && !(tty->flags & TTY_UTF8_ACS)) {
 		if (tty->term->acs[ch][0] == '\0')
 			return (NULL);
 		return (&tty->term->acs[ch][0]);
--- tty.c.orig	2015-09-27 18:51:49.000000000 +0900
+++ tty.c	2015-10-24 13:40:22.733122271 +0900
@@ -54,7 +54,7 @@
 void	tty_default_colours(struct grid_cell *, const struct window_pane *);
 
 #define tty_use_acs(tty) \
-	(tty_term_has((tty)->term, TTYC_ACSC) && !((tty)->flags & TTY_UTF8))
+	(tty_term_has((tty)->term, TTYC_ACSC) && !((tty)->flags & TTY_UTF8_ACS))
 
 #define tty_pane_full_width(tty, ctx) \
 	((ctx)->xoff == 0 && screen_size_x((ctx)->wp->screen) >= (tty)->sx)
